File: files/problems/longest-palindrome.c
int longestPalindrome(char* s) {
    int hash[128] = {0};
    int i = 0, count = 0;

    while (s[i]) {
        hash[s[i]]++;
        i++;
    }

    for (int i = 0; i < 128; i++) {
        if (hash[i] > 1) {
            if (hash[i] % 2 == 0)
                count += hash[i];
            else
                count += hash[i] - 1;
        }
    }

    for (int i = 0; i < 128; i++)
        if (hash[i] == 1 || hash[i] %2 == 1) 
            return count + 1;

    return count;
}

File: leetcode/problems/java/longest-palindrome.java
import java.util.HashMap;
import java.util.Map;

class Solution {
  public int longestPalindrome(String s) {
    Map<Character, Integer> map = new HashMap<>();
    for (char c : s.toCharArray()) {
      map.put(c, map.getOrDefault(c, 0) + 1);
    }

    int result = 0;
    boolean hasOdd = false;
    for (int count : map.values()) {
      if (count % 2 == 0) {
        result += count;
      } else {
        result += count - 1;
        hasOdd = true;
      }
    }

    return hasOdd ? result + 1 : result;
  }
}